### DISCUSSION TITLE
Usage with TRPC and Edge servers?

### DISCUSSION THREAD

**Discussion by b3nten** [digitallyinduced/thin-backend#60] [General]
Hi,

This project looks super interesting and I'm curious how it could work in a slightly different implementation.

Currently I'm enjoying using Prisma + TRPC + Next.js to build full-stack apps with end-to-end type safety. The downside is that Prisma can't run on the edge. I'm curious if it's possible to make queries using the Thin client on an edge server? If so combined with TRPC you'd have end-to-end type safety with edge functions, realtime on the client, and integrated auth. Basically my three main concerns when building apps.

Also what are the chances your auth implementation will support Twitch in the near (6 month) future?

Cheers,
Ben

**Comment by mpscholten** [comment]
Hey Ben,

Running Thin itself on the edge would work. The tricky part is the postgres. It's not easily possible to run postgres on the edge.

In the future we plan to have read replicas of the postgres db in multi regions. That way read latency to Thin is very low from everywhere. Write operations always need to go to the postgres master DB, so that has a bit of higher latency. But because of the optimistic UI updates for write operations, this latency is not visible at all.

There's a good chance for support Twitch auth 👍 Maybe you can open an issue for this so other people can +1 the issue if they're also interested in twitch login.

**Reply by b3nten** [reply]
Hi Marc, thanks for the response.

I'm not too concerned about postgres on the edge, but if the thin client can work in edge servers that's awesome. Basically I just want a way to call the DB on the server rather than on the client, but do so with type safety. Afaik no other solution lets you hit a postgres instance with typesafety from the edge.

I'll go make the issue for Twitch auth. Cheers!

-Ben

**Reply by mpscholten** [reply]
> I'm not too concerned about postgres on the edge, but if the thin client can work in edge servers that's awesome. Basically I just want a way to call the DB on the server rather than on the client, but do so with type safety. Afaik no other solution lets you hit a postgres instance with typesafety from the edge.

Can you share why you want to call the DB from the edge endpoint? Calling it from the client side has the advantage that the query can be a live query out of the box.

**Reply by b3nten** [reply]
My API endpoints do more than just fetch queries from the database. Ideally I could hit the DB with `service_role` permissions in order to basically do anything I want.

Basically I want to replace Prisma on the edge. I know thin.dev isn't supposed to be a Prisma alternative and I don't expect you to ever directly support my use-case. I'm more curious if if could be wrangled into it. Afaik there isn't anything else that can hit a postgres instance from the edge with type safety.

**Reply by mpscholten** [reply]
>  Ideally I could hit the DB with service_role permissions in order to basically do anything I want.

👍 

This is definitly planned for the future

Right now you could sort of emulate it by creating a specific user and then just hardcode the user_id of that service user into the policies.
